### PR TITLE
Add peer-to-peer connection support, and object exporting

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
This is needed to transition Ccokpit's internal dbus interface from it's current hacky/hardcoded implementation to something more reasonable.

### CODE REVIEW

**Comment by allisonkarlitskaya** [comment]
Had to kill some walruses.  Poor walruses. :(

**Comment by allisonkarlitskaya** [comment]
More RHEL missing stuff: `sd_bus_message_send()`, `random.randbytes()`, `asyncio.run()`...

**Comment by allisonkarlitskaya** [comment]
The centos crash is caused by the passing of the `sd_id128_t` to the `set_server()` call.  This is fixed in newer Python: https://bugs.python.org/issue22273

The calling convention for small structures is to put them into registers.  In this case, the 128 bit value would be treated equivalently to 2 64-bit values.  We could probably just do that, as a hack, instead.

**Comment by allisonkarlitskaya** [comment]
Directly passing two ints instead of a struct seemed pretty questionable, but I'm reasonably confident that a struct containing two (separate) ints is equivalent to a struct containing an array of two ints, even when passed-by-value.

File: systemd_ctypes/bus.py [hunk 1]
```diff
@@ -220,3 +223,125 @@ def add_match(self, rule, handler):
...
+    @classmethod
+    def prepare(cls):
+        raise NotImplementedError
+
+    def message_received(self, message):
+        try:
+            reply = message.new_method_return()
+            handled = self.handle_method_call(message, reply)
+        except BusError:
+            assert False  # needs proper exception handling
```
Comment (martinpitt) [inline systemd_ctypes/bus.py:244]: Let's add a few bad calls to the test? Nonexisting method, and method with mismatching signature?

File: systemd_ctypes/bus.py [hunk 2]
```diff
@@ -220,3 +223,125 @@ def add_match(self, rule, handler):
...
+    def append_property(self, reply, handler):
+        reply.open_container(ord('v'), handler._property_type)
+        reply.append_arg(handler._property_type, handler(self))
+        reply.close_container()
+
+    def handle_method_call(self, message, reply):
+        interface = message.get_interface(message)
+        method = message.get_member()
+
+        if interface == 'org.freedesktop.DBus.Introspectable':
```
Comment (martinpitt) [inline systemd_ctypes/bus.py:314]: Ah, I've missed The Big Switch™ since the days of TurboVision :grin:

**Review by martinpitt — COMMENTED** [review]
Niice! I have some little questions and remarks. LMK if you want me to add some test coverage! I can't do it today (hacking with Jelle still), but tomorrow on the train ride back. (I can also do that as a follow-up, though)

File: systemd_ctypes/bus.py [hunk 3]
```diff
@@ -159,6 +161,27 @@ def default_user(attach_event=False):
                 Bus._default_user.attach_event(None, 0)
         return Bus._default_user
 
+    @staticmethod
+    def from_socket(sock, server_id=None, attach_event=False):
+        self = Bus()
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:169]: I actually think that `self` is the most logical name for it.  This is, in effect a constructor.  The fact that Python makes this difficult is annoying, sure...

Do you have a suggestion for a different name?  `bus`?

File: systemd_ctypes/bus.py [hunk 4]
```diff
@@ -159,6 +161,27 @@ def default_user(attach_event=False):
...
+        sd.bus.new(self)
+        if server_id is not None:
+            self.set_server(True, server_id)
+        fd = os.dup(sock.fileno())
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:173]: Python's `socket` object owns the fd and will `close()` it when it goes away.  systemd also wants to own the fd.  I'll add a comment.

File: systemd_ctypes/bus.py [hunk 5]
```diff
@@ -31,6 +31,9 @@
 logger = logging.getLogger(__name__)
 
 class BusMessage(sd.bus_message):
+    def get_bus(self):
+        return Bus.ref(super().get_bus())
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:35]: It's two things:

 - `Message.get_bus()` returns a `sd.bus_p`, which is the lowlevel pointer type and doesn't have any of the API on it.  We need to create the highlevel wrapper for it here in order to be able to call methods on it.
 - `Message.get_bus()` returns a weak ref, and our highlevel wrapper assumes that it holds a strong ref, so indeed, we need to increase the refcount.

Some additional thinking at the layer of the librarywrapper could probably come up with a clever way to do this automatically, but it turns out that it's not a super common pattern.  The only other case that this occurs is the weak pointer to the `message` object passed into the message handler callback.  You can see that at line ~126:

```python
      def _callback(self, _message, _userdata, _ret_error):
          return 1 if self.__func__(BusMessage.ref(_message)) else 0
```

File: systemd_ctypes/bus.py (hunk 1 above)
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:244]: Sure.  I tested those manually, fwiw, and they did the right thing.  More tests welcome, of course :) :)

File: systemd_ctypes/bus.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:314]: naja...

You actually find the one spot where this approach suffers the most: we have to return a single XML document for introspection.  Multiple interfaces at the same object path could otherwise peacefully co-exist (and even properties would work correctly) but the XML is one document, generated in one spot.

Also TODO: I need to actually generate that document :rofl:

Comment (martinpitt) [reply systemd_ctypes/bus.py:314]: Don't worry, that was just a reminiscence of the past, not any suggestion to change it.

File: systemd_ctypes/bus.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply systemd_ctypes/bus.py:169]: Ah, I see -- I didn't look at this as a ctor.

File: systemd_ctypes/bus.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply systemd_ctypes/bus.py:173]: I found `socket.detach()` which is much nicer.

**Review by martinpitt — APPROVED** [review]
Nice! This is magic. :magic_wand:
